emailproxy.py

```diff
@@ -574,6 +574,25 @@ class OAuth2Helper:
     class TokenRefreshError(Exception):
         pass
 
+    @staticmethod
+    def get_cryptographer(config, username, password):
+        token_salt = config.get(username, 'token_salt', fallback=None)
+
+        # we hash locally-stored tokens with the given password
+        if not token_salt:
+            token_salt = base64.b64encode(os.urandom(16)).decode('utf-8')
+            config.set(username, 'token_salt', token_salt)
+
+        # generate encrypter/decrypter based on password and random salt
+        try:
+            decoded_salt = base64.b64decode(token_salt.encode('utf-8'))  # catch incorrect third-party proxy guide
+        except binascii.Error:
+            return ('%s: Invalid `token_salt` value found in config file entry for account %s - this value is not '
+                    'intended to be manually created; please remove and retry' % (APP_NAME, username))
```

Comment:
This could also be rewritten to stomp on previous values and just show a warning instead.

Reply:
[This](https://hkdb.medium.com/office365-via-oauth2-with-geary-1c0a1429226e) is the post in question that for whatever reason gave that incorrect advice and led to a few invalid issues being opened (and this code being added). Traffic seems to have subsided from that source recently, so yes, replacing the value with a new salt and just showing a warning would probably be fine. So, something like:

```python
        # we hash locally-stored tokens with the given password
        token_salt = config.get(username, 'token_salt', fallback=None)

        # generate encrypter/decrypter based on password and random salt
        try:
            decoded_salt = base64.b64decode(token_salt.encode('utf-8'))  # catch incorrect third-party proxy guide
        except binascii.Error:
            Log.info('%s: Invalid `token_salt` value found in config file entry for account %s - this value is not '
                    'intended to be manually created; please remove and retry' % (APP_NAME, username))
            token_salt = None

        if not token_salt:
            token_salt = base64.b64encode(os.urandom(16)).decode('utf-8')

        config.set(username, 'token_salt', token_salt)
```